plugin/pkg/managedseed/shoot/admission.go

```diff
@@ -0,0 +1,198 @@
+func (v *Shoot) ValidateInitialization() error {
+	if v.coreClient == nil {
+		return errors.New("missing garden core client")
+	}
+	if v.seedManagementClient == nil {
+		return errors.New("missing garden seedmanagement client")
+	}
+	return nil
+}
+
+var _ admission.ValidationInterface = &Shoot{}
+
+// Validate validates if the Shoot can be deleted. If the
+func (v *Shoot) Validate(ctx context.Context, a admission.Attributes, _ admission.ObjectInterfaces) error {
+	// Wait until the caches have been synced
+	if v.readyFunc == nil {
+		v.AssignReadyFunc(func() bool {
+			for _, readyFunc := range readyFuncs {
+				if !readyFunc() {
+					return false
+				}
+			}
+			return true
+		})
+	}
+	if !v.WaitForReady() {
+		return admission.NewForbidden(a, errors.New("not yet ready to handle request"))
+	}
+
+	// Ignore all kinds other than ManagedSeed
+	if a.GetKind().GroupKind() != seedmanagementv1alpha1.Kind("ManagedSeed") {
+		return nil
+	}
+
+	// Ignore updates to status or other subresources
+	if a.GetSubresource() != "" {
+		return nil
+	}
+
+	switch {
+	case a.GetName() == "":
+		return v.validateDeleteCollection(ctx, a)
+	default:
+		return v.validateDelete(ctx, a)
+	}
+}
+
+func (v *Shoot) validateDeleteCollection(ctx context.Context, a admission.Attributes) error {
+	managedSeeds, err := v.getManagedSeeds(ctx, labels.Everything())
+	if err != nil {
+		return err
+	}
+	for _, managedSeed := range managedSeeds {
+		if err := v.validateDelete(ctx, newAttributesWithName(a, managedSeed.Name)); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (v *Shoot) validateDelete(ctx context.Context, a admission.Attributes) error {
+	seedName := a.GetName()
+
+	shoots, err := v.getShoots(ctx, labels.Everything())
+	if err != nil {
+		return err
+	}
+
+	if admissionutils.IsSeedUsedByShoot(seedName, getShootPointersFromShoots(shoots)) {
+		return admission.NewForbidden(a, fmt.Errorf("cannot delete managed seed %s/%s since its seed %s is still used by shoot(s)", a.GetNamespace(), a.GetName(), a.GetName()))
+	}
+
+	return nil
+}
+
+func newAttributesWithName(a admission.Attributes, name string) admission.Attributes {
+	return admission.NewAttributesRecord(a.GetObject(),
+		a.GetOldObject(),
+		a.GetKind(),
+		a.GetNamespace(),
+		name,
+		a.GetResource(),
+		a.GetSubresource(),
+		a.GetOperation(),
+		a.GetOperationOptions(),
+		a.IsDryRun(),
+		a.GetUserInfo())
+}
+
+func (v *Shoot) getManagedSeeds(ctx context.Context, selector labels.Selector) ([]seedmanagementv1alpha1.ManagedSeed, error) {
+	managedSeedList, err := v.seedManagementClient.SeedmanagementV1alpha1().ManagedSeeds("").List(ctx, metav1.ListOptions{LabelSelector: selector.String()})
+	if err != nil {
+		return nil, err
+	}
+	return managedSeedList.Items, nil
+}
+
+func (v *Shoot) getShoots(ctx context.Context, selector labels.Selector) ([]core.Shoot, error) {
+	shootList, err := v.coreClient.Core().Shoots("").List(ctx, metav1.ListOptions{LabelSelector: selector.String()})
```

Comment:
This is quite the heavy call for both server and client. Although, it's not happening very frequently, I would like to either switch to a metadata-only list with a field selector or lister-based implementation.

Reply:
This is the same as https://github.com/gardener/gardener/blob/37db1f192d88482b1e36926979f5a1d0a20e170e/plugin/pkg/shoot/managedseed/admission.go#L173-L179 which was introduced as part of the ManagedSeed implementation (reviewed and approved by you back then). Is there any special reason to prefer a different implementation here?
A metadata-only list is not possible because we need `spec.seedName` and `status.seedName`. So the only option would be a lister-based implementation. But listers are cache-only and so we might miss an update to a shoot. I'm still quite confused how to properly decide between "cache-based, might be stale" vs. "client-based, so heavy".

Reply:
Done as suggested, in a new commit.
/squash

Reply:
I think, the difference is, that the `ShootManagedSeed` lists all shoots on a `DELETECOLLECTION shoots` call, whereas the `ManagedSeedShoot` admission plugin lists all shoots for each `{DELETE,DELETECOLLECTION} managedseeds` call.
metadata-only would be possible when only taking `spec.seedName` into account (can filter via a field selector).

> I'm still quite confused how to properly decide between "cache-based, might be stale" vs. "client-based, so heavy".

This is rather a gut feeling for me, there is no formal rule. Though, a good rule of thumb might be: avoid direct calls without filtering (i.e. no label selector, no field selector, especially across all namespaces) wherever possible. If operating on stale data from a lister would be really problematic, try some optimization for direct calls (e.g. filtering, metadata-only)

Reply:
OK, thanks for explaining. I think in this particular case cache misses are not that problematic (`spec.seedName` doesn't change too often), and so a lister-based implementation should be fine.